### PR TITLE
chore: merge the app with the container & implement the ApplicationContract

### PR DESCRIPTION
**Pull request by SychO9**
**Changes proposed in this pull request:**
Illuminate components always expect the app to be the container, but also expect the app to be implementing the laravel app contract. This means that very often between minor illuminate updates we get a call to a method on the app that doesn't exist in the Flarum app. This fixes the issue once and for all.

**Reviewers should focus on:**
The application contract implementation was separated into a trait to reduce the bloat because all those methods we don't actually need/use, so it can serve as a readable way of telling.

**Necessity**

- [x] Has the problem that is being solved here been clearly explained?
- [x] If applicable, have various options for solving this problem been considered?
- [x] For core PRs, does this need to be in core, or could it be in an extension?
- [x] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [ ] Frontend changes: tested on a local Flarum installation.
- [ ] Backend changes: tests are green (run `composer test`).
- [ ] Core developer confirmed locally this works as intended.
- [ ] Tests have been added, or are not appropriate here.

### CODE REVIEW

File: framework/core/src/Foundation/Config.php [hunk 1]
```diff
@@ -69,4 +69,9 @@ public function offsetUnset($offset): void
     {
         throw new RuntimeException('The Config is immutable');
     }
+
+    public function environment()
```
Comment (askvortsov1) [inline framework/core/src/Foundation/Config.php]: Is the lack of return type intentional?

File: php-packages/testing/src/integration/TestCase.php [hunk 2]
```diff
@@ -46,6 +46,8 @@ protected function tearDown(): void
     protected function app()
     {
         if (is_null($this->app)) {
+            $this->config('env', 'testing');
```
Comment (askvortsov1) [inline php-packages/testing/src/integration/TestCase.php:49]: Why was this added / why wasn't it needed before?

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+use Flarum\Locale\LocaleManager;
+use Illuminate\Support\Arr;
+use Illuminate\Support\Str;
+
+/**
+ * Implementation of the Laravel Application Contract,
+ * for the sake of better integration with Laravel packages/ecosystem.
+ */
+trait InteractsWithLaravel
```
Comment (askvortsov1) [inline framework/core/src/Foundation/Concerns/InteractsWithLaravel.php:20]: A lot of this stuff feels familiar, didn't we have it written somewhere before?

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        $this->terminatingCallbacks[] = $callback;
+
+        return $this;
+    }
+
+    public function terminate(): void
+    {
+        $index = 0;
+
+        while ($index < count($this->terminatingCallbacks)) {
```
Comment (askvortsov1) [inline framework/core/src/Foundation/Concerns/InteractsWithLaravel.php]: Why not a regular for loop?

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    {
+        return $this->joinPaths($this->paths->base, $path);
+    }
+
+    public function publicPath($path = ''): string
+    {
+        return $this->joinPaths($this->paths->public, $path);
+    }
+
+    public function storagePath($path = ''): string
```
Comment (askvortsov1) [inline framework/core/src/Foundation/Concerns/InteractsWithLaravel.php:48]: Don't we have this in `Paths`?

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        if (count($environments) > 0) {
+            $patterns = is_array($environments[0]) ? $environments[0] : $environments;
+
+            return Str::is($patterns, $this['env']);
+        }
+
+        return $this['env'];
+    }
+
+    public function runningInConsole(): bool
```
Comment (askvortsov1) [inline framework/core/src/Foundation/Concerns/InteractsWithLaravel.php:119]: Do we need to cache this?

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        return null;
+    }
+
+    public function isDownForMaintenance(): bool
+    {
+        // TODO: Implement isDownForMaintenance() method.
+        return false;
+    }
+
+    public function registerConfiguredProviders()
```
Comment (askvortsov1) [inline framework/core/src/Foundation/Concerns/InteractsWithLaravel.php]: What's up with the empty ones?

File: php-packages/testing/src/integration/TestCase.php (hunk 2 above)
Comment (SychO9) [reply php-packages/testing/src/integration/TestCase.php:49]: since we now implement the laravel `ApplicationContract` which has the  `environement` method that can be called internally in illuminate components, we might as well also make sure the value is correct in testing. (that's the gist of the changes in this PR)

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php (hunk 3 above)
Comment (SychO9) [reply framework/core/src/Foundation/Concerns/InteractsWithLaravel.php:20]: predates my time in flarum, but it might have implemented the laravel ApplicationContract at some point and then removed.

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php (hunk 5 above)
Comment (SychO9) [reply framework/core/src/Foundation/Concerns/InteractsWithLaravel.php:48]: yes, all these methods are implementations of the ApplicationContract, the method itself uses `Paths`

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php (hunk 6 above)
Comment (SychO9) [reply framework/core/src/Foundation/Concerns/InteractsWithLaravel.php:119]: it's cached

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php (hunk 7 above)
Comment (SychO9) [reply framework/core/src/Foundation/Concerns/InteractsWithLaravel.php]: has no actual meaning within our architecture, but have to implement the contract.

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php (hunk 6 above)
Comment (askvortsov1) [reply framework/core/src/Foundation/Concerns/InteractsWithLaravel.php:119]: Sorry, poor wording on my end. I mean, is this expensive to compute / is there a reason why we're caching it?

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php (hunk 7 above)
Comment (askvortsov1) [reply framework/core/src/Foundation/Concerns/InteractsWithLaravel.php]: Are any of the methods / code in this trait used for our needs at all?

Comment (askvortsov1) [reply framework/core/src/Foundation/Concerns/InteractsWithLaravel.php]: Also, could we mark all these as deprecated to discourage using them in Flarum code?

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php (hunk 6 above)
Comment (SychO9) [reply framework/core/src/Foundation/Concerns/InteractsWithLaravel.php:119]: oh, no it's not 🙈

File: framework/core/src/Foundation/Concerns/InteractsWithLaravel.php (hunk 7 above)
Comment (SychO9) [reply framework/core/src/Foundation/Concerns/InteractsWithLaravel.php]: > Are any of the methods / code in this trait used for our needs at all?

Indirectly, within illuminate components we use, like `runningInUnitTests` or `runningInConsole`, `terminating`. At any moment an Illuminate component could start calling one of these methods.
